qir/qat/Passes/GroupingPass/Tests/Unit/grouping.cpp

```diff
@@ -0,0 +1,525 @@
+
+TEST(GroupingComponent, ReuseTest)
+{
+    ConfigurationManager configuration_manager;
+    auto                 ir_manip = newIrManip(R"script(
+  %z = call i64 @get_arg()
+  %0 = icmp slt i64 %z, 0
+  %a1.i.i = select i1 %0, %Qubit* null, %Qubit* inttoptr (i64 1 to %Qubit*)
+  %a2.i.i = select i1 %0, %Qubit* inttoptr (i64 1 to %Qubit*), %Qubit* null
+  tail call void @__quantum__qis__x__body(%Qubit* %a1.i.i)
+  tail call void @__quantum__qis__z__body(%Qubit* null)
+  tail call void @__quantum__qis__cnot__body(%Qubit* %a1.i.i, %Qubit* %a2.i.i)
+  tail call void @__quantum__qis__mz__body(%Qubit* %a2.i.i, %Result* null)
+  tail call void @__quantum__qis__reset__body(%Qubit* %a2.i.i)
+  %1 = tail call i1 @__quantum__qis__read_result__body(%Result* null)
+  tail call void @__quantum__qis__x__body(%Qubit* %a1.i.i)
+  tail call void @__quantum__qis__z__body(%Qubit* null)
+  tail call void @__quantum__qis__cnot__body(%Qubit* %a1.i.i, %Qubit* %a2.i.i)
+  tail call void @__quantum__qis__mz__body(%Qubit* %a2.i.i, %Result* null)
+  tail call void @__quantum__qis__reset__body(%Qubit* %a2.i.i)
+  %2 = tail call i1 @__quantum__qis__read_result__body(%Result* null)
+  %3 = mul i64 %z, 45
+  %.op.op.i.i = add i64 %3, 13
+  %4 = select i1 %1, i64 -7, i64 %.op.op.i.i
+  %5 = select i1 %2, i64 -7, i64 %.op.op.i.i  
+  call void @no_opt(i64 %4)
+  call void @no_opt(i64 %5)
+  )script");
+
+    auto adaptor = newQirAdaptor(configuration_manager);
+    ir_manip->applyQirAdaptor(adaptor);
+
+    EXPECT_TRUE(ir_manip->hasInstructionSequence(
+        {
+            "%0 = call i64 @get_arg()",
+            "%1 = icmp slt i64 %0, 0",
+            "br label %load",
+
+        },
+        "entry"));
+
+    EXPECT_TRUE(ir_manip->hasInstructionSequence(
+        {
+            "%2 = select i1 %1, %Qubit* null, %Qubit* inttoptr (i64 1 to %Qubit*)",
+            "%3 = select i1 %1, %Qubit* inttoptr (i64 1 to %Qubit*), %Qubit* null",
+            "br label %quantum",
+
+        },
+        "load"));
+
+    EXPECT_TRUE(ir_manip->hasInstructionSequence(
+        {
+            "tail call void @__quantum__qis__x__body(%Qubit* %2)",
+            "tail call void @__quantum__qis__z__body(%Qubit* null)",
+            "tail call void @__quantum__qis__cnot__body(%Qubit* %2, %Qubit* %3)",
+            "br label %readout",
+        },
+        "quantum"));
+
+    EXPECT_TRUE(ir_manip->hasInstructionSequence(
+        {
+            "tail call void @__quantum__qis__mz__body(%Qubit* %3, %Result* null)",
+            "tail call void @__quantum__qis__reset__body(%Qubit* %3)",
+            "%4 = tail call i1 @__quantum__qis__read_result__body(%Result* null)",
+            "br label %post-classical",
+
+        },
+        "readout"));
+
+    EXPECT_TRUE(ir_manip->hasInstructionSequence(
+        {
+            "%5 = mul i64 %0, 45",
+            "%6 = add i64 %5, 13",
+            "%7 = select i1 %4, i64 -7, i64 %6",
+            "call void @no_opt(i64 %7)",
+            "br label %load4",
```

Comment:
I don't see what the load4 block is here. Is it empty? It looks like there might be two blocks of quantum instructions in this case due to the second set of quantum instructions acting on the same qubit as is measured before. I am curious how smart the dependence analysis is; what would happen if the two lines 
```
tail call void @__quantum__qis__z__body(%Qubit* null)
tail call void @__quantum__qis__cnot__body(%Qubit* %a1.i.i, %Qubit* %a2.i.i)
```
weren't there and only 
```
tail call void @__quantum__qis__x__body(%Qubit* %a1.i.i)
```
is there after a measurement of %Qubit* %a2.i.i?
Is the dependence analysis customized or is it built-in llvm determining that? I guess it needs to be custom, given how qubits and results are represented.

Reply:
Yes, `load4` is just an empty block. The dependency analysis is not that smart. If you remove the two instructions you suggest, the grouping pass will still split it into two quantum groups because of 
```
%3 = select i1 %1, %Qubit* inttoptr (i64 1 to %Qubit*), %Qubit* null
```
The analysis cannot determine whether measurement is affecting qubit zero or one, so it have to assume that the state is destroyed.

It is a cusomised dependency analysis. It could be extended to be smarter, but would require quite a bit more work.